### DISCUSSION TITLE
Share context

### DISCUSSION THREAD

**Discussion by Hromon** [dotnet/Silk.NET#534] [Q&A] [answered]
Hi o/
How do I share multiple contexts in one window?

**Comment by Perksey** [comment, marked as answer]
You can't have multiple contexts in one window. You can share one window's context with another window's context (i.e. so you can have one main window and one invisible background window to allow for multithreading) but multiple contexts in a single window isn't something we can support on every platform, thus it has been excluded from windowing.

<sub>If this answer is acceptable please use the "Mark as answer" button</sub>

**Reply by Hromon** [reply]
Ok, how i can share context between windows?

**Reply by Perksey** [reply]
An example could look like:
```cs
// create the first window
var window1 = Window.Create(WindowOptions.Default);

// use initialize to open the window - this is needed so we can share its context
window1.Initialize();

// create window 2 which shares the window's context
var window2opts = WindowOptions.Default;
window2opts.ShareContext = window1.GLContext;
var window2 = Window.Create(window2opts);

// open window 2
window2.Initialize();
```

You will then need to implement your own game loop so you can run Window 2's render thread on another thread, but keep running the events on the main thread (which is a requirement, you'll break some operating systems otherwise)

For your reference, here is the [standard single threaded game loop](https://github.com/dotnet/Silk.NET/blob/main/src/Windowing/Silk.NET.Windowing.Common/WindowExtensions.cs#L45-L60).
- **Initialize** creates the window.
- **DoEvents** must be ran on the main thread. That is, the thread that the program started on. This powers every event but Update and Render.
- **DoRender** and **DoUpdate** are the functions you want to run on another thread for Window 2.
- **Reset** closes the window.
- **IsClosing** is true when the close button is pressed or you use the `Close` method on the window.

**Reply by Hromon** [reply]
Thanks.
